src/backend/metal/src/command.rs

```diff
@@ -751,18 +754,25 @@ impl RawCommandQueue<Backend> for CommandQueue {
             if let Some(ref signal_block) = signal_block {
                 msg_send![command_buffer, addCompletedHandler: signal_block.deref() as *const _];
             }
-            // only append the fence handler to the last buffer
-            if i + 1 == num_buffers {
-                if let Some(ref fence) = fence {
-                    let value_ptr = fence.0.clone();
-                    let fence_block = ConcreteBlock::new(move |_cb: *mut ()| -> () {
-                        *value_ptr.lock().unwrap() = true;
-                    }).copy();
-                    msg_send![command_buffer, addCompletedHandler: fence_block.deref() as *const _];
-                }
+            if !inner.retained_buffers.is_empty() {
+                let retained_buffers = mem::replace(&mut inner.retained_buffers, Vec::new());
+                let release_block = ConcreteBlock::new(move |_cb: *mut ()| -> () {
+                    let _ = retained_buffers; // move and auto-release
+                });
+                msg_send![command_buffer, addCompletedHandler: release_block.deref() as *const _];
             }
             command_buffer.commit();
         }
+
+        if let Some(ref fence) = fence {
+            let command_buffer = self.0.new_command_buffer_ref();
+            let value_ptr = fence.0.clone();
+            let fence_block = ConcreteBlock::new(move |_cb: *mut ()| -> () {
+                *value_ptr.lock().unwrap() = true;
+            }).copy();
```

Comment:
Not part of your change, but why is the `copy` needed here?

Reply:
uh, seems totally out of place